torchaudio/datasets/tedlium.py

```diff
@@ -0,0 +1,158 @@
+        "checksum": "ad1e454d14d1ad550bc2564c462d87c7a7ec83d4dc2b9210f22ab4973b9eccdb",
+        "data_path": "data/",
+        "subset": None,
+    },
+}
+
+Tedlium_item = namedtuple(
+    "Tedlium_item", ["waveform", "sample_rate", "transcript", "talk_id", "speaker_id", "identifier"]
+)
+
+
+class TEDLIUM(Dataset):
+    """
+    Create a Dataset for Tedlium. Each item is a tuple of the form:
+    [waveform, sample_rate, transcript, talk_id, speaker_id, identifier]
+    """
+
+    def __init__(
+        self, root: str, release: str = RELEASE, subset: str = None, download: bool = False, audio_ext=".sph"
+    ) -> None:
+        """Constructor for TEDLIUM dataset
+
+        Args:
+            root (str): Path containing dataset or target path where its downloaded if needed
+            release (str, optional): TEDLIUM identifier (release1,release2,release3). Defaults to RELEASE.
+            subset (str, optional): Subset of data(train,test,dev) supported for release 1,2. Defaults to Train/None.
+            download (bool, optional): Download dataset in case is not founded in root path. Defaults to False.
+            audio_ext (str, optional): Overwrite audio extension when loading items. Defaults to ".sph".
+
+        Raises:
+            RuntimeError: If release identifier does not match any supported release,
+        """
+        self._ext_audio = audio_ext
+        if release in _RELEASE_CONFIGS.keys():
+            folder_in_archive = _RELEASE_CONFIGS[release]["folder_in_archive"]
+            url = _RELEASE_CONFIGS[release]["url"]
+            subset = subset if subset else _RELEASE_CONFIGS[release]["subset"]
+        else:
+            # Raise warning
+            raise RuntimeError(
+                "The release {} does not match any of the supported tedlium releases{} ".format(
+                    release, _RELEASE_CONFIGS.keys(),
+                )
+            )
+
+        basename = os.path.basename(url)
+        archive = os.path.join(root, basename)
+
+        basename = basename.split(".")[0]
+
+        self._path = os.path.join(root, folder_in_archive, _RELEASE_CONFIGS[release]["data_path"])
+        if subset in ["train", "dev", "test"]:
+            self._path = os.path.join(self._path, subset)
+        if download:
+            if not os.path.isdir(self._path):
+                if not os.path.isfile(archive):
+                    checksum = _RELEASE_CONFIGS[release]["checksum"]
+                    download_url(url, root, hash_value=checksum)
+                extract_archive(archive)
+
+        walker = walk_files(self._path, suffix=".stm", prefix=False, remove_suffix=True)
+        self._walker = list(walker)
+        self._extended_walker = []
+        for file in self._walker:
+            stm_path = os.path.join(self._path, "stm", file + ".stm")
+            with open(stm_path) as f:
+                l = len(f.readlines())
+                self._extended_walker += [(file, line) for line in range(l)]
```

Comment:
I checked the contents of STM files. For the case of `release-3`

```
$ wc -l TEDLIUM_release1/*/stm/*.stm | tail -1
   58863 total
$ wc -l TEDLIUM_release2/*/stm/*.stm | tail -1
   95033 total
$ wc -l TEDLIUM_release-3/data/stm/* | tail -1
  268263 total
```

So if loading the whole `release-3` dataset, `self._extended_walker` could take a couple of megabytes on memory.


Also this will open the number of files;

```
$ find  TEDLIUM_release1 -type f -name '*.stm' | wc -l
793
$ find  TEDLIUM_release2 -type f -name '*.stm' | wc -l
1514
$ find  TEDLIUM_release-3 -type f -name '*.stm' | wc -l
2370
```

For `release-3` dataset, this initialization process will go over 2370 files sequentially. How long does this take?

Reply:
oh is this what your report as 24 micros second??

Reply:
Yes, I report the time of initiating the dataset and per se going over all files counting lines. I think that is quite fast and we shouldn't worry about it.

About the size it occupies in memory, I did a quick test for a comment from @vincentqb just on top of this one. I checked how much would the walker occupy if we were to store the full line and not the line identifier as we already loop it, saving some I/O when loading each item.

I went ahead and checked the size of the walker list for release3 (the biggest)
```python
import numpy as np
np.array(test_dataset._extended_walker).nbytes
```
At its `316733528 bytes,~316 megabytes` storing the full line and `20449080 ~20 megabytes` with just the line identifier. I would say is a considerable size but it does remove a I/O operation each time we load the item. Does any other dataset take this much space in memory?

Reply:
> Yes, I report the time of initiating the dataset and per se going over all files counting lines. I think that is quite fast and we shouldn't worry about it.

I did a quick benchmark on my end and 

<details><summary>script</summary>

```python
import time

import torchaudio

torchaudio.set_audio_backend('sox_io')

n_rep = 100
t0 = time.monotonic()
for _ in range(n_rep):
    ds = torchaudio.datasets.TEDLIUM("../dataset/tedlium", release="release3", download=False)
t1 = time.monotonic()
print(len(ds._walker))
print((t1 - t0) / n_rep)
```

</details>

```
$python foo.py
268263
0.21009648161940275
```

I did not get as fast as micro seconds, but it was only 0.2 seconds so I agree that speed is not an issue here.

Reply:
Regarding the memory consumption, I talked with @cpuhrsch. 20 megabytes is not a big deal. However be aware that when using multiple data loader worker process, each process will consume the same amount of memory.

Also DataLoader has memory leakage issue https://github.com/pytorch/pytorch/issues/13246 .